### ISSUE TITLE
Feature: CORS Configuration in the UI / API

### ISSUE THREAD

**Issue by robotdan** [FusionAuth/fusionauth-issues#180]
### Problem
FusionAuth ships a CORS configuration, but there are many use cases and sometimes users bump into CORS issues that require support or modifications to our stock configuration. 

### Solution
Expose a CORS configuration in the user interface to enable easy modification of our CORS configuration at runtime. 

### Solution details

1. The user should be able to enable or disable the CORS filter in FusionAuth UI or via API
2. The user should be able to modify the following CORS parameters
  - Allowed origins, probably default to `*`. 
  - Allowed methods (? Maybe not necessary unless a user wishes to restrict this further than what we already enforce)
  - Allowed headers (? Maybe not necessary, TBD)
  - Exposed headers
3. Supports credentials (boolean)
4. Decorate request (boolean)
5. Optionally? Allow a configuration by URL pattern, this may or may not be exposed
  - For example `/api/*` for a positive filter, or a negative match such as `/admin/*` meaning do not apply the CORS filter to anything under `/admin/*`. 

We'll want to limit the times we return the `Access-Control-Allow-Origin: * ` header to API related endpoints such as `/api/*`. Requests to `/admin/*` should have a header of `Access-Control-Allow-Origin: https://foo.fusionauth.io`. (where `foo.fusionauth.io` is the current domain)
### Alternatives/workarounds
Modify the `web/WEB-INF/web.xml` on each node running FusionAuth and restart the web service. 👎 

### Additional context
- https://github.com/FusionAuth/fusionauth-issues/issues/44
- https://github.com/FusionAuth/fusionauth-issues/issues/105
- https://github.com/FusionAuth/fusionauth-issues/issues/171

### How to vote
Please give us a thumbs up or thumbs down as a reaction to help us prioritize this feature. Feel free to comment if you have a particular need or comment on how this feature should work.

**Comment by robotdan** [comment]
@AlvMF1 feel free to add any comments or feedback on this proposed feature.

**Comment by botnyx** [comment]
The workaround works for me.

But a more finegrained control over CORS would be appreceated.
i've needed the workaround to make calls to the well-known/openid configuration.

`Access to XMLHttpRequest at 'https://fusionauth/.well-known/openid-configuration/jwks' from origin 'https://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

**Comment by robotdan** [comment]
This feature is planned for the upcoming release, nearing code completion. We will allow CORS to be disabled, or enabled and once enabled modify the default configuration. 

The current shipped configuration does allow access to the JWKS endpoint via CORS. I wonder if your URL is incorrect? 

These two URLs should be going through the CORS filter and adding the `Access-Control-Allow-Origin` header. 

```
https://fusionauth/.well-known/jwks
https://fusionauth/.well-known/openid-configuration
```

**Comment by robotdan** [comment]
Coming soon!
https://fusionauth.io/docs/v1/tech/reference/cors

**Comment by dmushkov** [comment]
.well-known/openid-configuration still does not handled by cors filter

**Comment by mooreds** [comment]
@dmushkov please file a new issue with this info; unfortunately comments on closed issues often get lost.
